### ISSUE TITLE
v3.2.4 Hanging daemon templates in v3.2 when using `WorkflowTemplate`

### ISSUE THREAD

**Issue by sevberg** [argoproj/argo-workflows#7388]
### What happened/what you expected to happen?

Tried to use Argo Workflows v3.2.4 (upgrading from v3.1.14) on account of desirable new features in 3.2.4. Sadly, our production pipelines which make use of daemon templates began hanging indefinitely. Looking into it further showed that this is not isolated to our pipeline, but also effects the [daemon-step example](https://github.com/argoproj/argo-workflows/blob/master/examples/daemon-step.yaml) if it is re-implemented as a ClusterWorkflowTemplate (provided below). 

### What version is it broken in?
3.2.4

### What version was it working in?
3.1.14

### Reproduction

You should (hopefully) be able to reproduce the behavior if you install v3.2.4, apply the following ClusterWorkflowTemplate, and execute it with `argo -n <NAMESPACE> submit --from wftmpl/daemon-step` 

```yaml
---
apiVersion: argoproj.io/v1alpha1
kind: ClusterWorkflowTemplate
metadata:
  name: cluster-daemon-step
spec:
  entrypoint: daemon-example
  templates:
    - name: daemon-example
      steps:
        - - name: influx
            template: influxdb
        - - name: init-database
            template: influxdb-client
            arguments:
              parameters:
                - name: cmd
                  value: curl -XPOST 'http://{{steps.influx.ip}}:8086/query' --data-urlencode "q=CREATE DATABASE mydb"
        - - name: producer-1
            template: influxdb-client
            arguments:
              parameters:
                - name: cmd
                  value: for i in $(seq 1 20); do curl -XPOST 'http://{{steps.influx.ip}}:8086/write?db=mydb' -d "cpu,host=server01,region=uswest load=$i" ; sleep .5 ; done
          - name: producer-2
            template: influxdb-client
            arguments:
              parameters:
                - name: cmd
                  value: for i in $(seq 1 20); do curl -XPOST 'http://{{steps.influx.ip}}:8086/write?db=mydb' -d "cpu,host=server02,region=uswest load=$((RANDOM % 100))" ; sleep .5 ; done
          - name: producer-3
            template: influxdb-client
            arguments:
              parameters:
                - name: cmd
                  value: curl -XPOST 'http://{{steps.influx.ip}}:8086/write?db=mydb' -d 'cpu,host=server03,region=useast load=15.4'
        - - name: consumer
            template: influxdb-client
            arguments:
              parameters:
                - name: cmd
                  value: sleep 600
    - name: influxdb
      daemon: true
      container:
        image: influxdb:1.2
        readinessProbe:
          httpGet:
            path: /ping
            port: 8086
          initialDelaySeconds: 5
          timeoutSeconds: 1
    - name: influxdb-client
      inputs:
        parameters:
          - name: cmd
      container:
        image: appropriate/curl:latest
        command: ["sh", "-c"]
        args: ["{{inputs.parameters.cmd}}"]
---
apiVersion: argoproj.io/v1alpha1
kind: WorkflowTemplate
metadata:
  name: "daemon-step"
  namespace: "accure"
spec:
  entrypoint: wrapped-workflow
  templates:
    - name: wrapped-workflow
      steps:
        - - name: wrapped-daemon-example
            templateRef:
              name: "cluster-daemon-step"
              template: daemon-example
              clusterScope: true
```

### What Kubernetes provider are you using?
AWS EKS (v1.19.13)

### What executor are you running? Docker/K8SAPI/Kubelet/PNS/Emissary
docker

### Logs
```bash
# Logs from the workflow controller:
kubectl logs -n argo deploy/workflow-controller | grep ${workflow}

# If the workflow's pods have not been created, you can skip the rest of the diagnostics.

# The workflow's pods that are problematic:
kubectl get pod -o yaml -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded

# Logs from in your workflow's wait container, something like:
kubectl logs -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```

#### kubectl logs -n argo deploy/workflow-controller | grep ${workflow}
```text
time="2021-12-13T09:57:01.543Z" level=info msg="Processing workflow" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:01.579Z" level=info msg="Updated phase  -> Running" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:01.580Z" level=info msg="Steps node daemon-step-qf7xr initialized Running" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:01.580Z" level=info msg="StepGroup node daemon-step-qf7xr-1132653877 initialized Running" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:01.581Z" level=info msg="Steps node daemon-step-qf7xr-2458278934 initialized Running" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:01.581Z" level=info msg="StepGroup node daemon-step-qf7xr-2778624136 initialized Running" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:01.581Z" level=info msg="Pod node daemon-step-qf7xr-3198460218 initialized Pending" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:01.620Z" level=info msg="Created pod: daemon-step-qf7xr[0].wrapped-daemon-example[0].influx (daemon-step-qf7xr-3198460218)" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:01.620Z" level=info msg="Workflow step group node daemon-step-qf7xr-2778624136 not yet completed" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:01.620Z" level=info msg="Workflow step group node daemon-step-qf7xr-1132653877 not yet completed" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:01.620Z" level=info msg="TaskSet Reconciliation" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:01.620Z" level=info msg=reconcileAgentPod namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:01.648Z" level=info msg="Workflow update successful" namespace=accure phase=Running resourceVersion=38508888 workflow=daemon-step-qf7xr
time="2021-12-13T09:57:11.622Z" level=info msg="Processing workflow" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:11.622Z" level=info msg="Updating node daemon-step-qf7xr-3198460218 status Pending -> Running" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:11.624Z" level=info msg="Workflow step group node daemon-step-qf7xr-2778624136 not yet completed" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:11.624Z" level=info msg="Workflow step group node daemon-step-qf7xr-1132653877 not yet completed" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:11.624Z" level=info msg="TaskSet Reconciliation" namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:11.624Z" level=info msg=reconcileAgentPod namespace=accure workflow=daemon-step-qf7xr
time="2021-12-13T09:57:11.641Z" level=info msg="Workflow update successful" namespace=accure phase=Running resourceVersion=38508940 workflow=daemon-step-qf7xr
```
#### kubectl get pod -o yaml -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded
```text
apiVersion: v1
items:
- apiVersion: v1
  kind: Pod
  metadata:
    annotations:
      kubernetes.io/psp: eks.privileged
      workflows.argoproj.io/node-id: daemon-step-qf7xr-3198460218
      workflows.argoproj.io/node-name: daemon-step-qf7xr[0].wrapped-daemon-example[0].influx
    creationTimestamp: "2021-12-13T09:57:01Z"
    labels:
      workflows.argoproj.io/completed: "false"
      workflows.argoproj.io/workflow: daemon-step-qf7xr
    managedFields:
    - apiVersion: v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:workflows.argoproj.io/node-id: {}
            f:workflows.argoproj.io/node-name: {}
          f:labels:
            .: {}
            f:workflows.argoproj.io/completed: {}
            f:workflows.argoproj.io/workflow: {}
          f:ownerReferences:
            .: {}
            k:{"uid":"9eb3c0ee-4452-4ef6-8335-57bb178f7b0e"}:
              .: {}
              f:apiVersion: {}
              f:blockOwnerDeletion: {}
              f:controller: {}
              f:kind: {}
              f:name: {}
              f:uid: {}
        f:spec:
          f:containers:
            k:{"name":"main"}:
              .: {}
              f:env:
                .: {}
                k:{"name":"ARGO_CONTAINER_NAME"}:
                  .: {}
                  f:name: {}
                  f:value: {}
                k:{"name":"ARGO_DEADLINE"}:
                  .: {}
                  f:name: {}
                  f:value: {}
                k:{"name":"ARGO_INCLUDE_SCRIPT_OUTPUT"}:
                  .: {}
                  f:name: {}
                  f:value: {}
                k:{"name":"ARGO_TEMPLATE"}:
                  .: {}
                  f:name: {}
                  f:value: {}
              f:image: {}
              f:imagePullPolicy: {}
              f:name: {}
              f:readinessProbe:
                .: {}
                f:failureThreshold: {}
                f:httpGet:
                  .: {}
                  f:path: {}
                  f:port: {}
                  f:scheme: {}
                f:initialDelaySeconds: {}
                f:periodSeconds: {}
                f:successThreshold: {}
                f:timeoutSeconds: {}
              f:resources: {}
              f:terminationMessagePath: {}
              f:terminationMessagePolicy: {}
            k:{"name":"wait"}:
              .: {}
              f:command: {}
              f:env:
                .: {}
                k:{"name":"ARGO_CONTAINER_NAME"}:
                  .: {}
                  f:name: {}
                  f:value: {}
                k:{"name":"ARGO_CONTAINER_RUNTIME_EXECUTOR"}:
                  .: {}
                  f:name: {}
                k:{"name":"ARGO_DEADLINE"}:
                  .: {}
                  f:name: {}
                  f:value: {}
                k:{"name":"ARGO_INCLUDE_SCRIPT_OUTPUT"}:
                  .: {}
                  f:name: {}
                  f:value: {}
                k:{"name":"ARGO_POD_NAME"}:
                  .: {}
                  f:name: {}
                  f:valueFrom:
                    .: {}
                    f:fieldRef:
                      .: {}
                      f:apiVersion: {}
                      f:fieldPath: {}
                k:{"name":"ARGO_TEMPLATE"}:
                  .: {}
                  f:name: {}
                  f:value: {}
                k:{"name":"ARGO_WORKFLOW_NAME"}:
                  .: {}
                  f:name: {}
                  f:value: {}
                k:{"name":"GODEBUG"}:
                  .: {}
                  f:name: {}
                  f:value: {}
              f:image: {}
              f:imagePullPolicy: {}
              f:name: {}
              f:resources: {}
              f:terminationMessagePath: {}
              f:terminationMessagePolicy: {}
              f:volumeMounts:
                .: {}
                k:{"mountPath":"/var/run/docker.sock"}:
                  .: {}
                  f:mountPath: {}
                  f:name: {}
                  f:readOnly: {}
          f:dnsPolicy: {}
          f:enableServiceLinks: {}
          f:restartPolicy: {}
          f:schedulerName: {}
          f:securityContext: {}
          f:terminationGracePeriodSeconds: {}
          f:volumes:
            .: {}
            k:{"name":"docker-sock"}:
              .: {}
              f:hostPath:
                .: {}
                f:path: {}
                f:type: {}
              f:name: {}
      manager: workflow-controller
      operation: Update
      time: "2021-12-13T09:57:01Z"
    - apiVersion: v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          f:conditions:
            k:{"type":"ContainersReady"}:
              .: {}
              f:lastProbeTime: {}
              f:lastTransitionTime: {}
              f:status: {}
              f:type: {}
            k:{"type":"Initialized"}:
              .: {}
              f:lastProbeTime: {}
              f:lastTransitionTime: {}
              f:status: {}
              f:type: {}
            k:{"type":"Ready"}:
              .: {}
              f:lastProbeTime: {}
              f:lastTransitionTime: {}
              f:status: {}
              f:type: {}
          f:containerStatuses: {}
          f:hostIP: {}
          f:phase: {}
          f:podIP: {}
          f:podIPs:
            .: {}
            k:{"ip":"172.31.42.95"}:
              .: {}
              f:ip: {}
          f:startTime: {}
      manager: kubelet
      operation: Update
      time: "2021-12-13T09:57:10Z"
    name: daemon-step-qf7xr-3198460218
    namespace: accure
    ownerReferences:
    - apiVersion: argoproj.io/v1alpha1
      blockOwnerDeletion: true
      controller: true
      kind: Workflow
      name: daemon-step-qf7xr
      uid: 9eb3c0ee-4452-4ef6-8335-57bb178f7b0e
    resourceVersion: "38508933"
    selfLink: /api/v1/namespaces/accure/pods/daemon-step-qf7xr-3198460218
    uid: 116c3c1b-6353-43cb-89eb-cbc9b9bb8996
  spec:
    containers:
    - command:
      - argoexec
      - wait
      - --loglevel
      - info
      env:
      - name: ARGO_POD_NAME
        valueFrom:
          fieldRef:
            apiVersion: v1
            fieldPath: metadata.name
      - name: ARGO_CONTAINER_RUNTIME_EXECUTOR
      - name: GODEBUG
        value: x509ignoreCN=0
      - name: ARGO_WORKFLOW_NAME
        value: daemon-step-qf7xr
      - name: ARGO_CONTAINER_NAME
        value: wait
      - name: ARGO_TEMPLATE
        value: '{"name":"influxdb","inputs":{},"outputs":{},"metadata":{},"daemon":true,"container":{"name":"","image":"influxdb:1.2","resources":{},"readinessProbe":{"httpGet":{"path":"/ping","port":8086},"initialDelaySeconds":5,"timeoutSeconds":1}}}'
      - name: ARGO_INCLUDE_SCRIPT_OUTPUT
        value: "false"
      - name: ARGO_DEADLINE
        value: "0001-01-01T00:00:00Z"
      image: argoproj/argoexec:v3.2.4
      imagePullPolicy: IfNotPresent
      name: wait
      resources: {}
      terminationMessagePath: /dev/termination-log
      terminationMessagePolicy: File
      volumeMounts:
      - mountPath: /var/run/docker.sock
        name: docker-sock
        readOnly: true
      - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
        name: default-token-p5588
        readOnly: true
    - env:
      - name: ARGO_CONTAINER_NAME
        value: main
      - name: ARGO_TEMPLATE
        value: '{"name":"influxdb","inputs":{},"outputs":{},"metadata":{},"daemon":true,"container":{"name":"","image":"influxdb:1.2","resources":{},"readinessProbe":{"httpGet":{"path":"/ping","port":8086},"initialDelaySeconds":5,"timeoutSeconds":1}}}'
      - name: ARGO_INCLUDE_SCRIPT_OUTPUT
        value: "false"
      - name: ARGO_DEADLINE
        value: "0001-01-01T00:00:00Z"
      image: influxdb:1.2
      imagePullPolicy: IfNotPresent
      name: main
      readinessProbe:
        failureThreshold: 3
        httpGet:
          path: /ping
          port: 8086
          scheme: HTTP
        initialDelaySeconds: 5
        periodSeconds: 10
        successThreshold: 1
        timeoutSeconds: 1
      resources: {}
      terminationMessagePath: /dev/termination-log
      terminationMessagePolicy: File
      volumeMounts:
      - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
        name: default-token-p5588
        readOnly: true
    dnsPolicy: ClusterFirst
    enableServiceLinks: true
    nodeName: ip-172-31-46-30.eu-central-1.compute.internal
    preemptionPolicy: PreemptLowerPriority
    priority: 0
    restartPolicy: Never
    schedulerName: default-scheduler
    securityContext: {}
    serviceAccount: default
    serviceAccountName: default
    terminationGracePeriodSeconds: 30
    tolerations:
    - effect: NoExecute
      key: node.kubernetes.io/not-ready
      operator: Exists
      tolerationSeconds: 300
    - effect: NoExecute
      key: node.kubernetes.io/unreachable
      operator: Exists
      tolerationSeconds: 300
    volumes:
    - hostPath:
        path: /var/run/docker.sock
        type: Socket
      name: docker-sock
    - name: default-token-p5588
      secret:
        defaultMode: 420
        secretName: default-token-p5588
  status:
    conditions:
    - lastProbeTime: null
      lastTransitionTime: "2021-12-13T09:57:01Z"
      status: "True"
      type: Initialized
    - lastProbeTime: null
      lastTransitionTime: "2021-12-13T09:57:10Z"
      status: "True"
      type: Ready
    - lastProbeTime: null
      lastTransitionTime: "2021-12-13T09:57:10Z"
      status: "True"
      type: ContainersReady
    - lastProbeTime: null
      lastTransitionTime: "2021-12-13T09:57:01Z"
      status: "True"
      type: PodScheduled
    containerStatuses:
    - containerID: docker://1a8a5acf3c87a254444083dd1689903fca6f7dfc37f06f9d4f77daceab1088fd
      image: influxdb:1.2
      imageID: docker-pullable://influxdb@sha256:20459db7820d8cc435b555b32f58973340bd9b47ee8392cbac89c755ab4eb442
      lastState: {}
      name: main
      ready: true
      restartCount: 0
      started: true
      state:
        running:
          startedAt: "2021-12-13T09:57:02Z"
    - containerID: docker://bd1489b7413b102bd15af58784cb61d80916a1a77d03ab5ad85d5a645bda9913
      image: argoproj/argoexec:v3.2.4
      imageID: docker-pullable://argoproj/argoexec@sha256:73aa3d438337d6c87a51207918d97253ea8c64b3edfde02ef1a981e34d2bfa4a
      lastState: {}
      name: wait
      ready: true
      restartCount: 0
      started: true
      state:
        running:
          startedAt: "2021-12-13T09:57:02Z"
    hostIP: 172.31.46.30
    phase: Running
    podIP: 172.31.42.95
    podIPs:
    - ip: 172.31.42.95
    qosClass: BestEffort
    startTime: "2021-12-13T09:57:01Z"
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

#### kubectl logs -c wait -l workflows.argoproj.io/workflow=${workflow},workflow.argoproj.io/phase!=Succeeded

```text
time="2021-12-13T17:47:27.364Z" level=info msg="docker ps --all --no-trunc --format={{.Status}}|{{.Label \"io.kubernetes.container.name\"}}|{{.ID}}|{{.CreatedAt}} --filter=label=io.kubernetes.pod.namespace=accure --filter=label=io.kubernetes.pod.name=daemon-step-qf7xr-3198460218"
time="2021-12-13T17:47:27.388Z" level=info msg="listed containers" containers="map[main:{1a8a5acf3c87a254444083dd1689903fca6f7dfc37f06f9d4f77daceab1088fd Up {0 63774986222 <nil>}} wait:{bd1489b7413b102bd15af58784cb61d80916a1a77d03ab5ad85d5a645bda9913 Up {0 63774986222 <nil>}}]"
time="2021-12-13T17:47:37.388Z" level=info msg="docker ps --all --no-trunc --format={{.Status}}|{{.Label \"io.kubernetes.container.name\"}}|{{.ID}}|{{.CreatedAt}} --filter=label=io.kubernetes.pod.namespace=accure --filter=label=io.kubernetes.pod.name=daemon-step-qf7xr-3198460218"
time="2021-12-13T17:47:37.412Z" level=info msg="listed containers" containers="map[main:{1a8a5acf3c87a254444083dd1689903fca6f7dfc37f06f9d4f77daceab1088fd Up {0 63774986222 <nil>}} wait:{bd1489b7413b102bd15af58784cb61d80916a1a77d03ab5ad85d5a645bda9913 Up {0 63774986222 <nil>}}]"
time="2021-12-13T17:47:47.417Z" level=info msg="docker ps --all --no-trunc --format={{.Status}}|{{.Label \"io.kubernetes.container.name\"}}|{{.ID}}|{{.CreatedAt}} --filter=label=io.kubernetes.pod.namespace=accure --filter=label=io.kubernetes.pod.name=daemon-step-qf7xr-3198460218"
time="2021-12-13T17:47:47.440Z" level=info msg="listed containers" containers="map[main:{1a8a5acf3c87a254444083dd1689903fca6f7dfc37f06f9d4f77daceab1088fd Up {0 63774986222 <nil>}} wait:{bd1489b7413b102bd15af58784cb61d80916a1a77d03ab5ad85d5a645bda9913 Up {0 63774986222 <nil>}}]"
time="2021-12-13T17:47:57.445Z" level=info msg="docker ps --all --no-trunc --format={{.Status}}|{{.Label \"io.kubernetes.container.name\"}}|{{.ID}}|{{.CreatedAt}} --filter=label=io.kubernetes.pod.namespace=accure --filter=label=io.kubernetes.pod.name=daemon-step-qf7xr-3198460218"
time="2021-12-13T17:47:57.476Z" level=info msg="listed containers" containers="map[main:{1a8a5acf3c87a254444083dd1689903fca6f7dfc37f06f9d4f77daceab1088fd Up {0 63774986222 <nil>}} wait:{bd1489b7413b102bd15af58784cb61d80916a1a77d03ab5ad85d5a645bda9913 Up {0 63774986222 <nil>}}]"
time="2021-12-13T17:48:07.480Z" level=info msg="docker ps --all --no-trunc --format={{.Status}}|{{.Label \"io.kubernetes.container.name\"}}|{{.ID}}|{{.CreatedAt}} --filter=label=io.kubernetes.pod.namespace=accure --filter=label=io.kubernetes.pod.name=daemon-step-qf7xr-3198460218"
time="2021-12-13T17:48:07.508Z" level=info msg="listed containers" containers="map[main:{1a8a5acf3c87a254444083dd1689903fca6f7dfc37f06f9d4f77daceab1088fd Up {0 63774986222 <nil>}} wait:{bd1489b7413b102bd15af58784cb61d80916a1a77d03ab5ad85d5a645bda9913 Up {0 63774986222 <nil>}}]"
```

---

**Message from the maintainers**:

Impacted by this regression? Give it a 👍. We prioritise the issues with the most 👍.

**Comment by alexec** [comment]
Can you please share your `workflow-controller-configmap`?

**Comment by sevberg** [comment]
> Can you please share your `workflow-controller-configmap`?

Nothing wild going on here:

```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: workflow-controller-configmap
data:
  persistence: |
    connectionPool:
      maxIdleConns: 100
      maxOpenConns: 0
      connMaxLifetime: 0s
    nodeStatusOffLoad: true
    archive: true
    archiveTTL: 7d
    postgresql:
      host: <PULUMI-POSTGRES-HOST-URL>
      port: <PULUMI-POSTGRES-HOST-PORT>
      database: <PULUMI-POSTGRES-HOST-DATABASE>
      tableName: argo_workflows
      userNameSecret:
        name: argo-postgres-config
        key: username
      passwordSecret:
        name: argo-postgres-config
        key: password
```

**Comment by alexec** [comment]
Can you please add `containerRuntimeExecutor: emissary` or `containerRuntimeExecutor: pns`  to this?

**Comment by sevberg** [comment]
With the emissary executor I got:

```
STEP                           TEMPLATE                            PODNAME                       DURATION  MESSAGE
 ✖ daemon-step-955xk           wrapped-workflow                                                            child 'daemon-step-955xk-2677373404' failed                                                                                                                                                                                                                                                                     
 └───✖ wrapped-daemon-example  cluster-daemon-step/daemon-example                                          step group deemed errored due to child daemon-step-955xk[0].wrapped-daemon-example[0].influx error: when using the emissary executor you must either explicitly specify the command, or list the image's command in the index: https://argoproj.github.io/argo-workflows/workflow-executors/#emissary-emissary  
     └───⚠ influx              influxdb                            daemon-step-955xk-2288619344  0s        when using the emissary executor you must either explicitly specify the command, or list the image's command in the index: https://argoproj.github.io/argo-workflows/workflow-executors/#emissary-emissary                                                                                                      

This workflow does not have security context set. You can run your workflow pods more securely by setting it.
Learn more at https://argoproj.github.io/argo-workflows/workflow-pod-security-context/
```

Meanwhile, in the `pns` case, the hanging behavior was still there

In case it's helpful: In both instances, I'm still able to run the basic [daemon-step example](https://github.com/argoproj/argo-workflows/blob/master/examples/daemon-step.yaml)

**Comment by terrytangyuan** [comment]
Looks like that example is not compatible with emissary executor. If you are using emissary executor, you'll need to specify the command explicitly for your `influxdb` container.

**Comment by alexec** [comment]
@sevberg does this ONLY happen when using `WorkflowTemplate` or `ClusterWorkflowTemplate`?

**Comment by sevberg** [comment]
@alexec, I haven't tested it with a `WorkflowTemplate`. But my assumption from #7408 is that the issue would be present there, as well. 

However, yes, when using a `ClusterWorkflowTemplate` I see the hanging behavior on the daemon template, where as I do not see it when directly submitting a `Workflow` (which I've been testing by submitting the daemon-step example).

**Comment by alexec** [comment]
}
4d12bbfee fix(conttroller): Always set finishedAt dote. Fixes #6135 (#6139)
diff --git a/workflow/controller/operator.go b/workflow/controller/operator.go
index 8401b7189..ea3d3fda4 100644
--- a/workflow/controller/operator.go
+++ b/workflow/controller/operator.go
@@ -1185,14 +1185,7 @@ func (woc *wfOperationCtx) assessNodeStatus(pod *apiv1.Pod, node *wfv1.NodeStatu
 
        if node.Fulfilled() && node.FinishedAt.IsZero() {
                updated = true
-               if !node.IsDaemoned() {
-                       node.FinishedAt = getLatestFinishedAt(pod)
-               }
-               if node.FinishedAt.IsZero() {
-                       // If we get here, the container is daemoned so the
-                       // finishedAt might not have been set.
-                       node.FinishedAt = metav1.Time{Time: time.Now().UTC()}
-               }
+               node.FinishedAt = getLatestFinishedAt(pod)
                node.ResourcesDuration = resource.DurationForPod(pod)
        }
        if updated {
@@ -1228,18 +1221,13 @@ func (woc *wfOperationCtx) cleanUpPod(pod *apiv1.Pod, tmpl wfv1.Template) {
        }
 }

**Comment by alexec** [comment]
Should be fixed on `latest` and will be on v3.2.7
